### ISSUE TITLE
Reduce import time

### ISSUE THREAD

**Issue by nspope** [tskit-dev/tsdate#438]
The import time (due to numba AOT compilation) is driving me nuts-- I wonder if there's a way we can build an extension optionally on install (e.g. triggered by a flag to pip). Jerome and others have had bad experiences with numba extensions, but 30 to 60 seconds for import really slows down the debug cycle, and if extension-building is turned off by default I don't think it'd be an issue.

**Comment by hyanwong** [comment]
I agree it's a pain. I'll check with @jeromekelleher and @benjeffery if there is a sensible solution in this case.

**Comment by jeromekelleher** [comment]
One simple thing is to turn caching on/off using an environment variable. So you turn on caching for dev, and it's off by default: https://github.com/sgkit-dev/sgkit/blob/main/sgkit/accelerate.py

Would this help?

**Comment by hyanwong** [comment]
That sounds like a great suggestion. Thanks @jeromekelleher

**Comment by nspope** [comment]
That's a good suggestion @jeromekelleher and should be fine for development, thanks. I'm also pondering moving some of the core routines into a rust extension, in the long term.

**Comment by jeromekelleher** [comment]
> I'm also pondering moving some of the core routines into a rust extension, in the long term.

eek!

**Comment by nspope** [comment]
It's not that bad, I promise! What I'm trying to say is I'd like to move some of the low-level EP stuff into a compiled language extension, to save on import time, as this is just going to keep creeping up. Rust is great for numerical stuff and plugs in really nicely to Python.

**Comment by jeromekelleher** [comment]
Should definitely make caching an option for people who know what they're doing, 60s is horrendous.

If you paste in a profile I can comment on some possible ameliorations - that sounds much worse than it should be

**Comment by nspope** [comment]
> Do you want to implement the environment variable caching

I'll play around with it a bit then ping you both. It's annoying but not super urgent.

**Comment by hyanwong** [comment]
I tried the approach using in sgkit, in #441, and am getting lots of messages like this:

```
hypergeo.py:147: NumbaWarning: Cannot cache compiled function "_betaln" as it uses dynamic globals (such as ctypes pointers and large global arrays)
  @numba_jit("f8(f8, f8)")
```

**Comment by nspope** [comment]
Yeah, I ran into the same issue-- it turns out all the EP stuff can't be cached (and that's the bulk of the compilation). Although I could probably drop the reliance on Scipy's Cython library (which is the root of the problem), I think the better way to go is to pull the EP stuff into a low-level extension.

**Comment by hyanwong** [comment]
I think https://github.com/numba/numba/issues/6972 seems relevant. For instance, this comment:

> When code like this is compiled:
>
> 
> ```
> @njit
> def bar():
>   pass
> 
> @njit
> def foo():
>     bar()
> 
> foo()
> ```
> 
> The function foo() "sees" bar as a global and the machine code for foo refers to bar by address, that address cannot be cached as it will change next time the program is run and so caching is disabled for function foo.

**Comment by nspope** [comment]
Hm ... it's not going to be possible to refactor things so as to avoid calling njit'd functions inside other njit'd functions (nor do I want to do this).

**Comment by benjeffery** [comment]
Would doing something like:

```python
from numba import njit, types

fntype = types.FunctionType(types.void())

@njit(cache=True)
def bar():
    pass

@njit(types.void(fntype), cache=True)
def foo(func):
    func()

foo(bar)
```
Be a simple refactor that would enable caching?

**Comment by nspope** [comment]
Thanks @benjeffery -- I'm not quite understanding, would this work because the signature is explicit? If so, that's already the case in tsdate (we're using all explicit signatures) and the reason it won't cache is because I'm using some ctypes globals to interface with scipy's special functions library here: https://github.com/tskit-dev/tsdate/blob/3c08becf500dac3a422376126e70314064c360fe/tsdate/hypergeo.py#L44-L68

which results in
```
NumbaWarning: Cannot cache compiled function "_betaln" as it uses dynamic globals (such as ctypes pointers and large global arrays)
  @numba_jit("f8(f8, f8)")
# etc
```

though now that I look at it carefully, it might be possible to avoid using scipy altogether here depending on what parts of `math` are supported by numba.

**Comment by benjeffery** [comment]
The trick isn't the explicit signature, but passing the global as an argument to the function:

`def _betaln(gamma_func, p, q)`

The issue here is that numba needs to know the exact type signature for all arguments to a function. Using a global makes it an implicit argument of the function (technically a "cell" object in the function's closure). But unlike normal arguments, it's complicated for numba to watch for changes to the global's type.

Passing the global down from the nearest caller that is not a numba jit'ed should enable caching.

**Comment by jeromekelleher** [comment]
Yeah, globals are to be avoided with numba for a few reasons.

**Comment by nspope** [comment]
Got it-- thanks!

**Comment by hyanwong** [comment]
So do we want to go down this route, and try caching a few more of the functions defined in https://github.com/tskit-dev/tsdate/issues/438#issuecomment-2509897834? That might be enough for testing purposes?

**Comment by nspope** [comment]
Hmm.. I think we can solve 99% of the problem by replacing the scipy gammaln wrapper with `math.lgamma` (which now works inside numba). `erfinv` and `gammaincinv` are no longer used (these wrappers can be removed), and `gammainc` is only used in one place (so we could use Ben's workaround for this alone).
